lib/libc/minimal/source/stdlib/strtol.c

```diff
@@ -96,9 +96,9 @@ long strtol(const char *nptr, char **endptr, register int base)
 	cutoff /= (unsigned long)base;
 	for (acc = 0, any = 0;; c = *s++) {
 		if (isdigit((unsigned char)c) != 0) {
-			c -= '0';
+			c = (char)c - '0';
```

Comment:
I'm confused. In C, character literals have `int` representation.

To make things spicy, in C++, they have `char` representation.

However, since this is a `.c` file, `'0'` is and `int`. Also, since `c` is an `int` here, there is no casting required.

Is it possible that this was a false positive, or that the tool was incorrectly used in C++ mode?

Reply:
Hi, the non-compliance is genuine, because MISRA C defines and uses the essential type system in its guidelines, rather than the standard C type system. In this case, a character constant consisting of a single character has essential type character, and Rule 10.2 states that a subtraction involving essentially character types should not have the form "ess. signed - ess. char" (where essentially signed in this case corresponds to C's standard type int), but rather "ess. char - ess. signed" or "ess. char - ess.char" (which is what the change proposes), and in the latter case the result has the standard type (int in this case), which is then ok to be assigned to 'c', otherwise others guidelines concerning the essential types of assignments would be violated.